pkg/ccl/changefeedccl/shared_cache.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package changefeedccl
+
+import (
+	"context"
+	"sync"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+)
+
+type SharedCache interface {
+	FetchWithLock(ctx context.Context, key []byte, fn func() ([]byte, error)) ([]byte, error)
+}
+
+type jobInfoCache struct {
+	job_id int
+	db     isql.DB
+}
+
+func sharedCacheForJob(id catpb.JobID, db isql.DB) SharedCache {
+	return jobInfoCache{job_id: int(id), db: db}
+}
+
+func newSharedCacheForCore() SharedCache {
+	return &inMemoryCache{}
+}
+
+func (j jobInfoCache) FetchWithLock(ctx context.Context, key []byte, fn func() ([]byte, error)) (val []byte, err error) {
```

Comment:
Sorry for snooping on a draft PR but I feel obligated to ask if we can use or enhance the accessors in `info_storage.go`?  Our Write method on InfoStorage under the hood is a delete of the existing row corresponding to the job_id, info_key, and an insert of the new row. The reasons for this are mentioned here - https://github.com/cockroachdb/cockroach/pull/82638/files#diff-d65357c15cffdb009c1b3fd0f6456265d25b94758f326207c531b8f734b59164R328-R348. So if you think this pattern will benefit future users then maybe we can move it to a helper. I'll leave it to your judgment and I'm excited to see new ways we use this table!